File: couchbase-entity-api/src/main/java/kaufland/com/coachbasebinderapi/PersistenceConfig.java
package kaufland.com.coachbasebinderapi;


import java.util.HashMap;
import java.util.Map;

import kotlin.reflect.KClass;

public class PersistenceConfig {

    private static PersistenceConfig mInstance;

    private final Connector mConnector;

    private PersistenceConfig(Connector connector) {
        mConnector = connector;
    }

    public interface Connector {

        Map<KClass<?>, TypeConversion> getTypeConversions();

        Map<String, Object> getDocument(String id, String dbName);

        void deleteDocument(String id, String dbName) throws PersistenceException;

        void upsertDocument(Map<String, Object> document, String id, String dbName) throws PersistenceException;

    }

    public static void configure(Connector connector) {
        mInstance = new PersistenceConfig(connector);
    }

    public Connector getConnector() {
        if (mConnector == null) {
            throw new RuntimeException("no database connector configured.. call PersistenceConfig.configure(Connector");
        }

        return mConnector;
    }

    public static PersistenceConfig getInstance() {
        return mInstance;
    }
}

File: couchbase-entity-api/src/main/java/kaufland/com/coachbasebinderapi/PersistenceConfig.kt
package kaufland.com.coachbasebinderapi

import kotlin.reflect.KClass

object PersistenceConfig {
    private var mConnector: Connector? = null
    private var mSuspendingConnector: SuspendingConnector? = null

    interface Connector {
        val typeConversions: Map<KClass<*>, TypeConversion>
        fun getDocument(id: String, dbName: String): Map<String, Any>?
        fun getDocuments(ids: List<String>, dbName: String): List<Map<String, Any>?>
        fun queryDoc(dbName: String, queryParams: Map<String, Any>): List<Map<String, Any>>

        @Throws(PersistenceException::class)
        fun deleteDocument(id: String, dbName: String)

        @Throws(PersistenceException::class)
        fun upsertDocument(document: MutableMap<String, Any>, id: String?, dbName: String)
    }

    interface SuspendingConnector{

        val typeConversions: Map<KClass<*>, TypeConversion>

        suspend fun getDocument(id: String, dbName: String): Map<String, Any>?

        suspend fun getDocuments(ids: List<String>, dbName: String): List<Map<String, Any>>

        suspend fun queryDoc(dbName: String, queryParams: Map<String, Any>): List<Map<String, Any>>

        @Throws(PersistenceException::class)
        suspend fun deleteDocument(id: String, dbName: String)

        @Throws(PersistenceException::class)
        suspend fun upsertDocument(document: MutableMap<String, Any>, id: String?, dbName: String)
    }

    val connector: Connector
        get() {
            if (mConnector == null) {
                throw RuntimeException("no database connector configured.. call PersistenceConfig.configure(Connector)")
            }
            return mConnector!!
        }

    val suspendingConnector: SuspendingConnector
        get() {
            if (mSuspendingConnector == null) {
                throw RuntimeException("no database suspendingConnector configured.. call PersistenceConfig.configure(SuspendingConnector)")
            }
            return mSuspendingConnector!!
        }

    fun configure(connector: Connector) {
        mConnector = connector
    }

    fun configure(connector: SuspendingConnector) {
        mSuspendingConnector = connector
    }
}